container/multi-value-slice/multi_value_slice.go

```diff
@@ -0,0 +1,380 @@
+	s.sharedItems = items
+	s.individualItems = map[interfaces.Id]*MultiValue[V]{}
+	s.appendedItems = []*MultiValue[V]{}
+	s.cachedLengths = map[interfaces.Id]int{}
+}
+
+// Len returns the number of items for the input object.
+func (s *Slice[V, O]) Len(obj O) int {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	l, ok := s.cachedLengths[obj.Id()]
+	if !ok {
+		return len(s.sharedItems)
+	}
+	return l
+}
+
+// Copy copies items between the source and destination.
+func (s *Slice[V, O]) Copy(src O, dst O) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, item := range s.individualItems {
+		for _, v := range item.Values {
+			found := compareIds(v.ids, src.Id(), func(_ int) {
+				v.ids = append(v.ids, dst.Id())
+			})
+			if found {
+				break
+			}
+		}
+	}
+
+appendedLoop:
+	for _, item := range s.appendedItems {
+		found := false
+		for _, v := range item.Values {
+			found = compareIds(v.ids, src.Id(), func(_ int) {
+				v.ids = append(v.ids, dst.Id())
+			})
+			if found {
+				break
+			}
+		}
+		if !found {
+			// This is an optimization. If we didn't find an appended item at index i,
+			// then all larger indices don't have an appended item for the object either.
+			break appendedLoop
+		}
+	}
+
+	srcLen, ok := s.cachedLengths[src.Id()]
+	if ok {
+		s.cachedLengths[dst.Id()] = srcLen
+	}
+}
+
+// Value returns all items for the input object.
+func (s *Slice[V, O]) Value(obj O) []V {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	result := make([]V, len(s.sharedItems))
+	for i, item := range s.sharedItems {
+		ind, ok := s.individualItems[uint64(i)]
+		if !ok {
+			result[i] = item
+		} else {
+			found := false
+			for _, v := range ind.Values {
+				found = compareIds(v.ids, obj.Id(), func(_ int) {
+					result[i] = v.val
+				})
+				if found {
+					break
+				}
+			}
+			if !found {
+				result[i] = item
+			}
+		}
+	}
+
+	for _, item := range s.appendedItems {
+		found := false
+		for _, v := range item.Values {
+			found = compareIds(v.ids, obj.Id(), func(_ int) {
+				result = append(result, v.val)
+			})
+			if found {
+				break
+			}
+		}
+		if !found {
+			// This is an optimization. If we didn't find an appended item at index i,
+			// then all larger indices don't have an appended item for the object either.
+			return result
+		}
+	}
+
+	return result
+}
+
+// At returns the item at the requested index for the input object.
+// If the object has an individual value at that index, it will be returned. Otherwise the shared value will be returned.
+// If the object has an appended value at that index, it will be returned.
+func (s *Slice[V, O]) At(obj O, index uint64) (V, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if index >= uint64(len(s.sharedItems)+len(s.appendedItems)) {
+		var def V
+		return def, fmt.Errorf("index %d out of bounds", index)
+	}
+
+	isOriginal := index < uint64(len(s.sharedItems))
+	if isOriginal {
+		ind, ok := s.individualItems[index]
+		if !ok {
+			return s.sharedItems[index], nil
```

Comment:
What if it's also not in the shared items mapping?

Reply:
We already check this here:
```
isOriginal := index < uint64(len(s.sharedItems))
```

If the index is within the length of shared items, it means we either have a shared value or an individual value. There's no third option.